### ISSUE TITLE
Question on loss calculation

### ISSUE THREAD

**Issue by staceycy** [j96w/DenseFusion#7]
Hi，<br>
Thank you for sharing the codes and I have several questions on the **loss calculation** :

Q1.
https://github.com/j96w/DenseFusion/blob/42a21e76ff5391d3703d4e144a0831faf6fbdf43/lib/loss.py#L58 Why do you combine one point from point clouds with pred_t here? 

Q2.
https://github.com/j96w/DenseFusion/blob/42a21e76ff5391d3703d4e144a0831faf6fbdf43/lib/loss.py#L67 I am sorry that I am confused about the reason of updating with this method. <br>Why the point cloud is updated by ****subtracting pred_t** before rotation**? Shouldn't it be updated the same as prediction(**adding pred_t after rotation**) ? 
>Following is some of my understanding:
1) To predict a residual pose, we can: <br> 1) update the model points same as prediction, while keeping the target the same; <br>2) or keep the model points the same an update the target in a reverse way. <br>

In the codes, the point clouds and target are updated in the same way, and it's hard for me to understand. Could you please help to explain? 

Thank you and looking forward to your reply. 

Best,
Stacey

**Comment by j96w** [comment]
Reply to Q1: Here the translation estimation is an offset translation. Since we are performing a per-pixel pose estimation, the translation estimation from each input point is an offset translation starts from the original 3D location of the point. This is how we add difference to the results from each input point and the network is trained unsupervisedly to choose which one should be the best prediction.

Reply to Q2: This line is preparing the ground truth pose of the object model for the next iterative refinement process. Since we already acquire the first estimation (ori_base(rotation) and ori_t(translation)) from the initial DenseFusion model, we can use this predicted pose inversely to make the ground truth model close to the origin of the camera frame. That's why it's subtracting. At the same time, the input point cloud(generated from the depth) should also apply the same inverse pose transformation and be used as the input for the next iteration. Your understanding is talking about the model points transformed by the predicted pose and the target, while this part of code is working on the input point cloud(generated from the depth) and the target for the next iteration, NOT the model points transformed by the predicted pose.

**Comment by prachees** [comment]
Hi, why is the translation estimation done as an offset (with respect to 3d points) vs estimating an absolute value? What advantages are observed?

**Comment by j96w** [comment]
We empirically found it's more robust than direct regress the absolute translation.

**Comment by sunshantong** [comment]
Hi, I am still confused about the new_target. Since target = model_points * target_t + target_t and _new_target_ = torch.bmm((new_target - ori_t), ori_base).contiguous(). So the inverse solution yields a similar result to model_points. Does it mean that _new_target_  belong to the world coordinate system just like model_points, not the camera coordinate system? So how we use  the _new_target_ as  the ground truth for the next iteration?

Thank you and looking forward to your reply.

**Comment by j96w** [comment]
Do not understand the result of the inverse solution as something related to the model_points. They are fundamentally different things. `tools/eval_ycb.py` or `tools/eval_linemod.py` will give you the answers.

**Comment by j96w** [comment]
Do not understand the result of the inverse solution as something similar to the model_points. They are fundamentally different things. `tools/eval_ycb.py` or `tools/eval_linemod.py` will give you the answers.

**Comment by sunshantong** [comment]
Thank you for your help. Sorry to bother you again. I still have some questions. 
1.The max default number of epochs to train is 500. When training 104 epochs on ycb, the _dis_ is 0.0090381440936. But it costs long time. And the ycb 
 trained_models you provide is `pose_refine_model_69_0.009449292959118935.pth`. So how to determine the epoch of training for pose refine model? The lower the dis, the better the model? How to prevent overfitting?

2.When evaluation on LineMOD Dataset, the final content of eval_result_logs.txt is as follows:
`No.13390 NOT Pass! Distance: 0.0217275395989
No.13391 Pass! Distance: 0.0048154219985
No.13392 Pass! Distance: 0.0142814125866
No.13393 Pass! Distance: 0.00356977432966
No.13394 Pass! Distance: 0.00472941761836
No.13395 Pass! Distance: 0.00784354563802
No.13396 Pass! Distance: 0.0127922594547
No.13397 Pass! Distance: 0.00581285078079
No.13398 NOT Pass! Distance: 0.0267377775162
No.13399 Pass! Distance: 0.00628646928817
No.13400 Pass! Distance: 0.0146940667182
No.13401 NOT Pass! Distance: 0.0340396799147
No.13402 NOT Pass! Distance: 0.0713591426611
No.13403 NOT Pass! Distance: 0.0522820688784
No.13404 Pass! Distance: 0.0038491380401
No.13405 NOT Pass! Distance: 0.0213586390018
No.13406 Pass! Distance: 0.00927203428
Object 1 success rate: 0
Object 2 success rate: 0
Object 4 success rate: 0
Object 5 success rate: 0
Object 6 success rate: 0
Object 8 success rate: 0
Object 9 success rate: 0
Object 10 success rate: 0
Object 11 success rate: 1
Object 12 success rate: 0
Object 13 success rate: 0
Object 14 success rate: 0
Object 15 success rate: 0
ALL success rate: 0`
Is anything wrong with evaluation on LineMOD Dataset？How can I get an evaluation result similar to YCB_Video Dataset run with MATLAB?

 Thank you in advance.

**Comment by j96w** [comment]
Hi @sunshantong, please don't worry, I will definitely try my best to help you solve your problem. But before that, could you open a new issue for your question? Seems like currently it's under the wrong topic?

**Comment by hillaric** [comment]
Did you understand what the author said, the eval_linemd.py has given you the answers?

**Comment by qingnanli** [comment]
Hi @j96w I have two questions on the loss. Could you please help me? 

Q1. In the line 98 of eval_linemod.py, pred_r, pred_t = refiner(new_points, emb, idx). Does the pred_r and pred_t mean the residual rotation and translation? If so, in loss_refiner.py, the target = updated new_target of the last iteration, and the pred = residual rotation pred_r  * model_points + resudual translation pred_t. There is no [R|T] = [Rk|Tk]...[R0|T0] like in loss_refiner.py (but exists in eval_linemod.py my_mat_final = np.dot(my_mat, my_mat_2)).

Q2. I am confused about new_target = torch.bmm((new_target - ori_t), ori_base).contiguous() in loss.py. At first, the new_target = ori_target[0].view(1, num_point_mesh, 3).contiguous(), where ori_target[0] is from dataset, and ori_t = points + offset pred_t. Therefore, (new_target - ori_t) means the points [1, 500, 3] (new_target - ori_t) are around the center of the coordinate. So when we apply torch.bmm((new_target - ori_t), ori_base), what does the updated new_target stand for?
